official/mnist/mnist.py

```diff
@@ -122,16 +127,45 @@ def model_fn(features, labels, mode, params):
         })
 
 
+def get_batch_size_multi(current_size):
+  """For multi-gpu, batch-size must be a multiple of the number of
+  available GPUs.
+
+  Note that this should eventually be handled by replicate_model_fn
+  directly. Multi-GPU support is currently experimental, however,
+  so doing the work here until that feature is in place.
+  """
+  from tensorflow.python.client import device_lib
+
+  local_device_protos = device_lib.list_local_devices()
+  num_gpus = sum([1 for d in local_device_protos if d.device_type == 'GPU'])
+  remainder = current_size % num_gpus
+  return current_size - remainder
+
+
 def main(unused_argv):
+  model_function = model_fn
+
+  if FLAGS.multi_gpu:
+    # Adjust batch size
+    FLAGS.batch_size = get_batch_size_multi(FLAGS.batch_size)
```

Comment:
Would it make sense to warn on changes (since it does mean that we may not completely respect the user's batch size).
Options:

1.  Make this:
    ```python
    batch_size = get_batch_size_multi(FLAGS.batch_size)
    if batch_size != FLAGS.batch_size:
      print(<something>)
      FLAGS.batch_size = batch_size
      ```

2. OR, maybe it's better to make this an assertion so that we always respect the user's `--batch_size` and fail when not? So instead of `get_batch_size_multi` we could have an `assert_batch_size_is_multiple_of_number_of_gpus` (I'm sure a better name is possible :)

The more I think about it, I think I prefer the latter (failing instead of silently changing). That way we never have to explain to someone why they don't see any change when they try `--batch_size=64` and `--batch_size=69` :)

Reply:
I agree with warning-- is printing the preferred method in this case? As to halting-- I'm fine if that's generally preferred, but my preference would be that it just does the right thing by default if I don't specify. Having to fully specify everything when all you want to do is run/try certain things can be frustrating. For example-- as a user, I just want to run this as fast as possible on whatever set of GPUs I have available. If we error out and request a batch size, I then have to go figure out how to do that.

Reply:
The exception message should be something helpful like: `Set --batch_size=128 or --batch_size=136 . --batch_size=129 does not evenly divide amongst the 8 GPUs`, telling the user exactly what they can do.

As for "does the right thing by default" - I agree. The question is whether the "right" thing is to "run however you can, possibly ignoring what the user asked for", or is the "right" thing to "run what the user specified".

Reply:
I guess in the ideal we would adjust if the user didn't specify at all and error if they specified something wrong, but that's probably more fine-tuning of configuration than is called for here given that this will eventually be handled by replicate_model_fn anyhow. In which case, an informative error message sounds good to me.